*: add protoutil.Unmarshal

petermattis opened the pull request:
Switch all remaining uses of proto.Message to protoutil.Message.

Changed the TestProtoUnmarshal lint to check for usage of
protoutil.Unmarshal instead of proto.Unmarshal. Added a TestProtoMessage
lint that checks for usage of proto.Message (bad) vs
protoutil.Message (good).

petermattis committed:
*: add protoutil.Unmarshal

Switch all remaining uses of proto.Message to protoutil.Message.

Changed the TestProtoUnmarshal lint to check for usage of
protoutil.Unmarshal instead of proto.Unmarshal. Added a TestProtoMessage
lint that checks for usage of proto.Message (bad) vs
protoutil.Message (good).

The performance improvement to BenchmarkValueGetProto is due to avoiding
a type assertion in proto.Unmarshal that the message implements an
Unmarshal method.

name             old time/op  new time/op  delta
ValueGetProto-8  30.8ns ± 2%  20.1ns ± 1%  -34.78%  (p=0.000 n=10+8)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18954)
<!-- Reviewable:end -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod a few import nits. Feel free to ignore

---

Reviewed 55 of 55 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/roachpb/api.go, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18954#-KvTcYXSo3k_t4EbVY1x:-KvTcYXSo3k_t4EbVY1y:bvypwbk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bacf3557a813e0c25ee007e48b801d1fc3d2a337/pkg/roachpb/api.go#L21)):*
> ```Go
> 	"strconv"
> 
> 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
> ```

Split CRDB include

---

*[pkg/security/auth.go, line 20 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18954#-KvTcc7VvY2k2IX3D_rn:-KvTcc7VvY2k2IX3D_ro:bb73xfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/bacf3557a813e0c25ee007e48b801d1fc3d2a337/pkg/security/auth.go#L20)):*
> ```Go
> 	"crypto/tls"
> 
> 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
> ```

Same

---

*[pkg/sql/drop_test.go, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18954#-KvTcoRK4bO3pfeBWI8P:-KvTcoRK4bO3pfeBWI8Q:bcloakl) ([raw file](https://github.com/cockroachdb/cockroach/blob/bacf3557a813e0c25ee007e48b801d1fc3d2a337/pkg/sql/drop_test.go#L37)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
> 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
> 	"github.com/pkg/errors"
> ```

Move up

---

*[pkg/sql/jobs/jobs.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18954#-KvTcqD5eQsuAYhD5vu-:-KvTcqD5eQsuAYhD5vu0:bcloakl) ([raw file](https://github.com/cockroachdb/cockroach/blob/bacf3557a813e0c25ee007e48b801d1fc3d2a337/pkg/sql/jobs/jobs.go#L36)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
> 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
> 	"github.com/pkg/errors"
> ```

Move up

---

*[pkg/util/httputil/http.go, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18954#-KvTdDvmUC5G8vuHGpdj:-KvTdDvmUC5G8vuHGpdk:b-80wwyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/bacf3557a813e0c25ee007e48b801d1fc3d2a337/pkg/util/httputil/http.go#L23)):*
> ```Go
> 	"strconv"
> 
> 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
> ```

Split

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18954#-:-KvTdVNAIEeEGKbJa2Gb:bw3uutm)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/api.go, line 21 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18954#-KvTcYXSo3k_t4EbVY1x:-KvTf5Qo9-_t_ZbvSO9U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bacf3557a813e0c25ee007e48b801d1fc3d2a337/pkg/roachpb/api.go#L21)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Split CRDB include
</blockquote></details>

Done.

---

*[pkg/security/auth.go, line 20 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18954#-KvTcc7VvY2k2IX3D_rn:-KvTfBVbJqCqNxV2uCCX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bacf3557a813e0c25ee007e48b801d1fc3d2a337/pkg/security/auth.go#L20)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same
</blockquote></details>

Done.

---

*[pkg/sql/drop_test.go, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18954#-KvTcoRK4bO3pfeBWI8P:-KvTfGX_ZSHVdhqAJ5qS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bacf3557a813e0c25ee007e48b801d1fc3d2a337/pkg/sql/drop_test.go#L37)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Move up
</blockquote></details>

Done.

---

*[pkg/sql/jobs/jobs.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18954#-KvTcqD5eQsuAYhD5vu-:-KvTfJTRfOAP028JfV6O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bacf3557a813e0c25ee007e48b801d1fc3d2a337/pkg/sql/jobs/jobs.go#L36)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Move up
</blockquote></details>

Done.

---

*[pkg/util/httputil/http.go, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18954#-KvTdDvmUC5G8vuHGpdj:-KvTfN70gmpmQE90n-wz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bacf3557a813e0c25ee007e48b801d1fc3d2a337/pkg/util/httputil/http.go#L23)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Split
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18954)*
<!-- Sent from Reviewable.io -->

petermattis commented:
This PR has the side-effect of speeding up `BenchmarkValueGetProto` due to avoiding a type assertion in `proto.Unmarshal` that the message implements an `Unmarshal` method.

```
name             old time/op  new time/op  delta
ValueGetProto-8  30.8ns ± 2%  20.1ns ± 1%  -34.78%  (p=0.000 n=10+8)
```
